### ISSUE TITLE
Ctrl+Shift+1 doesn't toggle system power as advertised

### ISSUE THREAD

**Issue by ianparkinson** [mjoblin/vibinui#283]
The help dialog suggests that ^⇧1 should toggle system power. It doesn't.

**Comment by ianparkinson** [comment]
I believe this to be a limitation of Mantine's `useHotkeys` system.

The implementation in `KeyboardShortcutsManager.tsx` adds an action to the shortcut `ctrl+shift+1`. Mantine won't match against this, because it expects the `1` to match the actual character typed - and `shift-1` types an exclamation mark `!`.

* `ctrl+shift+!` works, but is dependent on keyboard layout.
* `ctrl+1` also works, ignoring the shift key.

This is similar to the issue we ran into in #274, where it was difficult to trigger correctly on the `-`/`+` keys for similar reasons.

**Comment by mjoblin** [comment]
Interesting! Yeah this must be a keyboard layout thing like you mention. The existing hotkey works for me, but I only have a laptop keyboard and a separate tenkeyless, both set to US EN.

The reasons Vibin currently uses `ctrl+shift+1` are very me-specific: I find `ctrl+shift` easily pressable with just my thumb, and the `1` just felt convenient. It's also a key combination which is less likely to already be managed by an OS setting (e.g. I already have an unrelated `ctrl+1` hotkey defined at the OS level).

I did try `ctrl+shift+!` but that didn't work for me (likely given my keyboard layout, as you suggested).

I'm definitely open to changing how this works going forward. It's a shame the modal currently lies to anyone with an unsupported keyboard layout.

Some potential (non-exhaustive) options:

1. Replace the power toggle hotkey with a new one with fewer (ideally no) conflicts. Maybe `ctrl-shift-P`? (`P` for "Power"). This one works on my keyboard and I'd hope it works on others too). Definitely open to other suggestions here.
1. Consider replacement options for `useHotkeys`.
1. Support a full-blown user-defined hotkey system (although that's a big project with debatable payoff).

I don't feel strongly about keeping `ctrl-shift-1`, and I'm fine with keeping the work required to a minimum, so the first option is fine with me (if we do anything at all).

**Comment by ianparkinson** [comment]
Interesting, I'm quite surprised that it works for you - I must be misunderstanding something. Both US and UK keyboards should have the same behaviour for the `1` key, as far as I can tell.

If I set my system's keyboard layout to `English (US)`, it still doesn't trigger for me. Neither does it trigger if I use my laptop's inbuilt tenkeyless keyboard instead of my full-size keyboard. Perhaps a browser/OS difference? I'm on Chrome/Ubuntu.

Mantine's matching logic reads (slightly tweaked for readability):

```
    event.key.toLowerCase() === key.toLowerCase() ||
      event.code.replace('Key', '').toLowerCase() === key.toLowerCase()
```

...so it tries to match against both `event.key` and `event.code`, albeit with the `event.code` slightly mangled. `event.key` gives the character typed, `event.code` is the name of the pressed key.

I added a simple event listener to Chrome's console:

`document.addEventListener('keydown', (event) => { console.log(event.key, event.code) })`

..and then typed 1 and ctrl-shift-1 to get the `key` and `code` in each case. It logs:

```
1 Digit1
Shift ShiftLeft
Control ControlLeft
! Digit1
```

So the `event.key` is `1` or `!` depending on the shift state; and the code for the `1` key is `Digit1` regardless of modifier keys.

Maybe your system reports different values for ctrl-shift-1?

For me, the hotkey string `ctrl+shift+Digit1` does work as expected (or `ctrl+shift+Numpad1` for the other `1` key). I'm not sure whether either of those are intended to be supported by Mantine though, or whether they would work on your system.

**Comment by mjoblin** [comment]
Thanks for researching that!

I applied your event listener and got this with `1`:
```
1 Digit1
```

And this with `ctrl-shift-1`:
```
Control ControlLeft
Shift ShiftLeft
1 Digit1
```

... so basically the same as you, except for the `ctrl-shift-1` case where I get `1 Digit1` when you get `! Digit1`. Which aligns with the differences in behaviour we're seeing.

I tested `ctrl+shift+Digit1` on my setup and it worked fine (with both laptop keyboard and external keyboard). I'd be fine with switching over to that, and also adding support for `ctrl+shift+Numpad1` if you'd like to see that in there.

I'll test `ctrl+shift+Digit1` and `ctrl+shift+1` on Windows soon too, to compare. And I run Vibin on an Ubuntu Pi, which is normally headless, but I'll plug a monitor in and fire up Chrome and see what happens. Hopefully `Digit1` will work on all platforms. I'll report back.

(This all assumes that the current power toggle hotkey is desirable. I'm still open to hearing about other preferences, and I'm happy to switch to something else).

**Comment by mjoblin** [comment]
My tests up until now have been on macOS. Windows (a desktop not laptop, with just an external tenkeyless set to English (United States)) behaves as follows:

The event listener produces the following for `1`:
```
1 Digit1
```

And this with `ctrl-shift-1`:

```
Control ControlLeft
Shift ShiftLeft
! Digit1
```

This is the case in both Chrome and Firefox.

So Windows appears to behave like Ubuntu, with macOS being the outlier.

Changing `KeyboardShortcutsManager.tsx` from `ctrl+chift+1` to `ctrl+shift+Digit1` worked as expected on Windows (both browsers), which is good to confirm.

I still plan to test my Ubuntu/Pi setup soon.

**Comment by mjoblin** [comment]
I got the same result in Ubuntu 24.04.1 on a Raspberry Pi (I only tested in Firefox as the Pi is ARM and a non-chromebook ARM Chrome wasn't trivially available; but the following matches what you're seeing in Ubuntu/Chrome).

The event listener produces the following for `1`:

```
1 Digit1
```

And this with `ctrl-shift-1`:

```
Control ControlLeft
Shift ShiftLeft
! Digit1
```

Switching Vibin to `ctrl+shift+Digit1` also worked successfully on this setup.

If there's no alternatives to consider then I propose we switch `KeyboardShortcutsManager.tsx` to use `ctrl+shift+Digit1` as it seems to work fine on all tested platforms & browsers. We can also support `ctrl+shift+Numpad1` if you think that would be useful.

**Comment by ianparkinson** [comment]
Thanks for testing this so exhaustively. Interesting that Mac has a different behaviour - I guess it treats `ctrl+shift` as not shifting the character.

I've sent you #285 with that fix - it covers both `Numpad1` and `Digit1` so that either `1` key works.

I also looked at the hotkey specification you have for showing the help dialog: `shift+?`. I've confirmed that the hotkey spec string `?` *doesn't* work (Mantine expects the modifiers to match precisely, and the only way to type a `?` requires shifting). Fortunately, all [keyboard layouts that I can find](https://en.wikipedia.org/wiki/List_of_QWERTY_keyboard_language_variants) seem to have `?` bound to a shifted keypress so it should work pretty much universally.

**Comment by mjoblin** [comment]
Fantastic, thank you!

And thanks for investigating `shift+?` too. That one does seem a bit weird. It's curious that `?` doesn't work on its own.

I did a quick experiment trying to get `[` and `{` to display the help. On my keyboard, `[` and `{` are on the same key (with `[` being the default and `{` requiring Shift). This produced the same outcome as `?`, where `useHotkeys` worked with `[` to display the help, but requires `shift+{` when using `{` to display the help. So I guess it may be consistent for these dual-purpose keys. (I only tried this on Mac though).

And that Wikipedia page on keyboard layouts is fascinating. This has definitely been an adventure.
